storage/src/main/java/tech/pegasys/teku/storage/store/Store.java

```diff
@@ -390,21 +379,49 @@ public BeaconState getBlockState(Bytes32 blockRoot) {
 
   @Override
   public BeaconState getCheckpointState(Checkpoint checkpoint) {
+    return getCheckpointStateIfAvailable(checkpoint)
+        .orElseGet(
+            () -> {
+              final BeaconState checkpointState =
+                  regenerateCheckpointState(checkpoint, this::getBlockState);
+              lock.writeLock().lock();
+              try {
+                checkpoint_states.put(checkpoint, checkpointState);
+              } finally {
+                lock.writeLock().unlock();
+              }
+              return checkpointState;
+            });
+  }
+
+  private Optional<BeaconState> getCheckpointStateIfAvailable(final Checkpoint checkpoint) {
     readLock.lock();
     try {
-      return checkpoint_states.get(checkpoint);
+      final BeaconState state = checkpoint_states.get(checkpoint);
+      if (state != null) {
+        checkpointStateRequestCachedCounter.inc();
+        return Optional.of(state);
+      } else {
+        checkpointStateRequestMissCounter.inc();
+        return Optional.empty();
+      }
     } finally {
       readLock.unlock();
     }
   }
 
-  @Override
-  public boolean containsCheckpointState(Checkpoint checkpoint) {
-    readLock.lock();
+  private BeaconState regenerateCheckpointState(
+      final Checkpoint checkpoint, Function<Bytes32, BeaconState> getBlockState) {
     try {
-      return checkpoint_states.containsKey(checkpoint);
-    } finally {
-      readLock.unlock();
+      final BeaconState baseState = getBlockState.apply(checkpoint.getRoot());
+      if (baseState == null || baseState.getSlot().equals(checkpoint.getEpochStartSlot())) {
```

Comment:
Is this supposed to be:
```suggestion
      if (baseState != null || baseState.getSlot().equals(checkpoint.getEpochStartSlot())) {
```
I'm not sure why you would return a null pointer and then try to store that in checkpoint_states.

Reply:
Actually I do see why you would check for null and return here, but in that scenario, we still have to deal with the null object in the caller method.

Reply:
Good point, I've switched `regenerateCheckpointState` to return an `Optional` and flowed that around.